File: run_all_examples.py
# get ./src/examples all directories
import os
import glob
import typing

def build_moon_run_cmd(path : str ):
    # 	moon run src/examples/01-close/ --target native
    return f'moon run {path} --target native'

def clean_example_env():
    os.system('rm -rf ./tmp')

def build_example_env():
    os.system('mkdir -p ./tmp')
    with open('./tmp/input.txt', 'w') as f:
        f.write('Lorem ipsum odor amet, consectetuer adipiscing elit.')


def run_moon_run_cmd(path : str):
    clean_example_env()
    build_example_env()
    print(f"running: {os.path.basename(path)}")
    status = os.system(build_moon_run_cmd(path))
    
    if status != 0:
        exit(status)
        
    # if ./tmp/output.txt exists, print it
    out = "./tmp/output.txt"
    if os.path.exists(out):
        with open(out, 'r') as f:
            print(f"{out}: ")
            print(f.read())

def exclude_files(dir : str, exclude : list[str]):
    for basename in exclude:
        if dir.endswith(basename):
            return False 
    return True 

# main function
def main():
    clean_example_env()
    build_example_env()

    dirs = glob.glob('./src/examples/*')
    exclude = ['00-template', '35-pause']
    dirs = [dir for dir in dirs if exclude_files(dir,exclude) and os.path.isdir(dir)]
    # dir exclude and isdir

    dirs.sort()
    
    for dir in dirs:
        run_moon_run_cmd(dir)
        print("")

if __name__ == '__main__':
    main()

File: run_all_examples.ts
import * as fs from 'node:fs';
import cp from "node:child_process";
import { exit } from 'node:process';

let clean_example_env = () => {
  try {
    fs.rmSync("tmp", {
      recursive: true,
      force: true
    })
  } catch {
    // ignore errors
  }
}

let lorem = "Lorem ipsum odor amet, consectetuer adipiscing elit."
let build_example_env = () => {
  try {
    fs.mkdirSync("tmp")
    let fd = fs.openSync("tmp/input.txt", "w")
    fs.writeSync(fd, lorem)
    fs.closeSync(fd)
  } catch {
    // ignore errors
  }
}

let exe_path = (name: string): string => {
  return `./target/native/release/build/examples/${name}/${name}.exe`
}
let main = () => {
  // cp.execSync("moon build --target native")
  let progs = fs.readdirSync("src/examples", {
    withFileTypes: true
  }).filter((x) => x.name !== "35-pause")

  clean_example_env()

  for (let prog of progs) {
    build_example_env()
    try {
      console.log(`running: ${prog.name}`)
      let output = cp.execSync(exe_path(prog.name))
      console.log(output.toString())
    }
    catch {
      console.log("execute error")
      exit(-1)
    }
    clean_example_env()
  }
}
main()